File: 1-working-with-list/p02_last_two.ml

let rec last_two l =
  match l with
  | [] -> None
  | [x] -> None
  | [x; y] -> Some (x, y)
  | _ :: tl -> last_two tl;;


last_two [ "a" ; "b" ; "c" ; "d" ];;
last_two [ "a" ; "b" ];;
last_two [ "a" ];;
last_two [ ];;

let rec last_two' = function
  | [] | [_] -> None
  | [x; y] -> Some (x, y)
  | _ :: tl -> last_two tl;;

last_two' [ "a" ; "b" ; "c" ; "d" ];;
last_two' [ "a" ; "b" ];;
last_two' [ "a" ];;
last_two' [ ];;

File: 1-working-with-list/p02_last_two.scala
object Main {

  def main(args: Array[String]) {
    println(lastTwo(List("a", "b", "c", "d")))
    println(lastTwo(List("a")))
  }

  def lastTwo[T](l: List[T]): Option[(T, T)] = l match {
    case Nil => None
    case a :: Nil => None
    case a :: b :: Nil => Some((a, b))
    case hd :: rest => lastTwo(rest)
  }
}
